docs/spec/lightclient/attacks/evidence-handling.md

```diff
@@ -0,0 +1,250 @@
+
+# Light client attacks
+
+We define a light client attack as detection of conflicting headers for a given height that can be verified
+starting from the trusted light block. A light client attack is defined in the context of interactions of 
+light client with two peers. One of the peers (called primary) defines a trace of verified light blocks 
+(primary trace) that are being checked against trace of the other peer (called witness) that we call 
+witness trace. 
+
+A light client attack is defined by the primary and witness traces 
+that have a common root (the same trusted light block for a common height) but forms 
+conflicting branches (end of traces is for the same height but with different headers). 
+Note that conflicting branches could be arbitrarily big as branches continue to diverge after 
+a bifurcation point. We propose an approach that allows us to define a valid light client attack
+only with a common light block and a single conflicting light block. We rely on the fact that 
+we assume that the primary is under suspicion (therefore not trusted) and that the witness plays 
+support role to detect and process an attack (therefore trusted). Therefore, once a light client 
+detects an attack, it needs to send to a witness only missing data (common height
+and conflicting light block) as it has its trace. Keeping light client attack data of constant size 
+saves bandwidth and reduces an attack surface. As we will explain below, although in the context of 
+light client core 
+[verification](https://github.com/informalsystems/tendermint-rs/tree/master/docs/spec/lightclient/verification) 
+the roles of primary and witness are clearly defined, 
+in case of the attack, we run the same attack detection procedure twice where the roles are swapped. 
+The rationale is that the light client does not know what peer is correct (on a right main branch) 
+so it tries to create and submit an attack evidence to both peers. 
+   
+Light client attack evidence consists of a conflicting light block, a common height and an attack type. There are 
+three types of attacks that can be executed against Tendermint light client: 
+    - lunatic attack
+    - equivocation attack and 
+    - amnesia attack. 
+
+TODO: Add references to attack definitions.     
+TODO: Add Callum's figures.
+    
+```go
+type LightClientAttackEvidence struct {
+    ConflictingBlock   LightBlock
+    CommonHeight       int64
+    Type               AttackType
```

Comment:
Also I wonder if we should not add peerID of the node to which we are going to submit LightClientAttackEvidence to. At least internally. The point is that the evidence is specific to the peer for which we generate it.

Reply:
Given light client attack, there seems to be a simple way to differentiate between attack types. Keep in mind that the goal is to detect and punish +1/3 faulty processes. In case of lunatic attack (invalid block), we don't need to check for any other attack type as we have +1/3 of processes that precommitted invalid block. In case both headers are for valid blocks, we first check for pure equivocation attack. This is the case when commit round is the same. In this case we have enough information to punish +1/3 of equivocating processes. If commit round is different, we need to run full fork accountability protocol. In this case we could have mix of attacks: amnesia + equivocation, but we always have amnesia. Therefore we call such attack amnesia attack. Does this answer your question?

Reply:
I guess lunatic is clear.
Regarding the other two I am not sure. Even if both blocks are from round 5, it might still be that some faulty nodes did amnesia, and other faulty nodes did equivocation. Do I miss something?

Reply:
In case of pure equivocation attack the following holds: conflicting blocks are valid and commit round is the same. Note that the as conflicting blocks are valid that implies that validator sets are the same. Therefore, in order to have two commits, there need to be conflicting signatures of at least +1/3 validators. We could have amnesia attack in addition, but we don't need to detect it as we catch at least +1/3 of equivocations.

Reply:
Right!! Good point.  
I guess then we should just make it clear in the final write-up what precisely we are required to catch. I guess for every attack we have to catch at least *f + 1* faulty voting power. Perhaps we should also mention that in the future we might try to catch more, e.g., these additional amnesia scenarios.

Reply:
moreover, can't we deduce the attack type from ConflictingBlock?

Reply:
You need trusted block and conflicting block and then you can figure out the nature of attack. If this is your question. Common height is needed as part of attack evidence verification.